Спринт 12

IlyaKorastylev opened the pull request:
1. На экран поиска добавлен функционал сохранения и очистки истории поиска.

2. На экране настроек реализован переключатель тёмной темы.

IlyaKorastylev committed:
1. На экран поиска добавлен функционал сохранения и очистки истории поиска.
2. На экране настроек реализован переключатель тёмной темы.

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -82,12 +94,25 @@ class SearchActivity : AppCompatActivity() {
+        searchHistoryClearButton.setOnClickListener{
+            history.clear()
+            historyAdapter.notifyDataSetChanged()
+            searchHistoryLayout.visibility = if (history.trackList.size > 0) View.VISIBLE else View.GONE
```
Можно было бы использовать метод списка .isNotEmpty()

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -47,6 +43,11 @@ class SearchActivity : AppCompatActivity() {
+    private lateinit var searchHistoryClearButton: Button
+    private lateinit var history: SearchHistory
 
     override fun onCreate(savedInstanceState: Bundle?) {
```
Метод onCreate стал очень большим, можно было бы его разделить логически на несколько малых методов

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/SearchHistory.kt:
```diff
@@ -0,0 +1,34 @@
+const val SEARCH_HISTORY_KEY = "search_history"
+
+class SearchHistory(val sharedPrefs : SharedPreferences) {
+    var trackList : MutableList<Track> = getTracks().toMutableList()
```
Чаще всего для того, чтобы не допускать какие-то модификации извне, используют [backing properties](https://kotlinlang.org/docs/properties.html#backing-properties). В твоем случае было бы:  
```
 private val _trackList : MutableList<Track> = getTracks().toMutableList()
   
val trackList : List<Track> = _trackList
```

Снаружи используешь второй, внутри первый. Также метод getTracks можно сделать приватным

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/SearchHistory.kt:
```diff
@@ -0,0 +1,34 @@
+        sharedPrefs.edit()
+            .putString(SEARCH_HISTORY_KEY, json)
+            .apply()
+        }
```
Немного съехало форматирование

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/TrackAdapter.kt:
```diff
@@ -5,17 +5,29 @@ import androidx.recyclerview.widget.RecyclerView
-            return TrackViewHolder(parent)
+    internal var tracks = mutableListOf<Track>()
+    var onItemClick: ((Track) -> Unit)? = null
+    fun setTracks(newTracks: List<Track>?) {
```
Этот метод не используется, можно было бы сделать, как я написала выше в SearchHistory

Anna-Malinina commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -129,4 +131,53 @@
+            android:layout_height="wrap_content"
+            android:gravity="center"
+            android:layout_marginStart="16dp"
+            android:layout_marginEnd="16dp"
```
Можно использовать marginHorizontal, а значение вынести в ресурсы, если повторяется еще где-то

Anna-Malinina commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -129,4 +131,53 @@
+            android:fontFamily="@font/ys_display_medium"
+            android:textColor="?attr/colorOnPrimary"
+            android:text="@string/search_history_hint"
+            android:textSize = "19sp"
```
Лучше вынести в ресурсы

Anna-Malinina commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -129,4 +131,53 @@
+            android:layout_height="0dp"
+            android:layout_weight="1"
+            android:orientation="vertical"
+            android:paddingStart="13dp"
```
В дизайне могут быть нечетные размеры, или не симметричные. Но лучше использовать в коде размеры, кратные 4 или 8. Документация - https://material.io/design/layout/spacing-methods.html#baseline-grid

IlyaKorastylev committed:
1. В SearchActivity.kt в методе onTextChanged скорректировано условие видимости searchHistoryLayout для случая, если список истории пустой.
2. Создан общий стиль для кнопок на экране поиска.
3. Удален неиспользуемый метод setTracks.
